### PR TITLE
Run Swarm containers in provisioning step using Docker API instead of SSH/shell

### PR DESCRIPTION
**Pull request by kunalkushwaha**
Replaces the code of provisioing with Go Templates with DockerAPI

Signed-off-by: Kunal Kushwaha <kushwaha_kunal_v7@lab.ntt.co.jp>

### CODE REVIEW

**Comment by dgageot** [comment]
Hi @kunalkushwaha, I like the idea of this PR. Replacing templates with actual code seems so better. Looks like you need to `make fmt` though.

**Comment by kunalkushwaha** [comment]
Thanks,  It should be fine now.

**Comment by dgageot** [comment]
Thanks @kunalkushwaha, now it looks like there's a merge conflict. Maybe due to this commit https://github.com/docker/machine/commit/8f6b8da3999744594f0f3f8293e86491b21689db

**Comment by kunalkushwaha** [comment]
@dgageot Please review now once. 

BTW, I have integrated https://github.com/docker/machine/pull/2515 in my initial PR.

**Comment by dgageot** [comment]
@kunalkushwaha There's still a merge conflict

**Comment by kunalkushwaha** [comment]
@dgageot Sorry for conflict error. Forgot to fetch from upstream.

Now it should be fine.

**Comment by dgageot** [comment]
@kunalkushwaha fmt? :-)

**Comment by kunalkushwaha** [comment]
I saw that error on CI. But fmt report okay on my machine.

Is is ``go vet``? Vet reports lot issue with machine code.

Regards,
Kunal Kushwaha

------------------------------

**Comment by dgageot** [comment]
```
mk/validate.mk:13: recipe for target 'fmt' failed
```

**Comment by nathanleclaire** [comment]
Nice! I also really like this idea.

**Comment by kunalkushwaha** [comment]
Thanks :)

The error is really strange. On my machine ``go fmt`` reports okay, but ``make validate`` reports error.

```
~/go/src/github.com/docker/machine on  refactor-provisioning! ⌚ 9:27:57
$ go fmt libmachine/provision/configure_swarm.go 

~/go/src/github.com/docker/machine on  refactor-provisioning! ⌚ 9:28:20
$ make validate                                 
Congratulations! All commits are properly signed with the DCO!
libmachine/provision/configure_swarm.go
make: *** [fmt] Error 1

```

**Comment by nathanleclaire** [comment]
@kunalkushwaha You on Go 1.5.1 locally?

**Comment by kunalkushwaha** [comment]
Yes, Is this issue with Go 1.5.1 version?

**Comment by nathanleclaire** [comment]
@kunalkushwaha No, Go 1.5.1 is the desired version.  Are you running `gofmt -s -w **/*.go` in your shell?

**Comment by kunalkushwaha** [comment]
``gofmt -s -w **/*.go`` worked :) 

```
~/go/src/github.com/docker/machine on  refactor-provisioning! ⌚ 9:58:11
$ go version  
go version go1.5.1 linux/amd64

~/go/src/github.com/docker/machine on  refactor-provisioning! ⌚ 9:58:16
$ gofmt -s -w **/*.go

~/go/src/github.com/docker/machine on  refactor-provisioning! ⌚ 10:02:12
$ make validate                                                                                                                                                                                                                               
Congratulations! All commits are properly signed with the DCO!
go build \
        -o /home/kunal/go/src/github.com/docker/machine/bin/docker-machine \
         -tags "" \
        -ldflags "-X `go list ./version`.GitCommit=`git rev-parse --short HEAD 2>/dev/null` -w -s"  cmd/machine.go
libmachine/provision/configure_swarm.go:60: github.com/samalba/dockerclient.RestartPolicy composite literal uses unkeyed fields
libmachine/provision/configure_swarm.go:85: github.com/samalba/dockerclient.RestartPolicy composite literal uses unkeyed fields
exit status 1
make: *** [vet] Error 1

```

``` patch
--- a/libmachine/provision/configure_swarm.go
+++ b/libmachine/provision/configure_swarm.go
@@ -59,7 +59,7 @@ func configureSwarm(p Provisioner, swarmOptions swarm.Options, authOptions auth.
                masterHostConfig := dockerclient.HostConfig{
                        RestartPolicy: dockerclient.RestartPolicy{"Always", 0},
                        Binds:         []string{hostBind},
-                       PortBindings:  map[string][]dockerclient.PortBinding{"3376/tcp": []dockerclient.PortBinding{{"", port}}},
+                       PortBindings:  map[string][]dockerclient.PortBinding{"3376/tcp": {{"", port}}},
                        NetworkMode:   "host",
                }
 

```

Still vet reports error.

File: libmachine/provision/configure_swarm.go [hunk 1]
```diff
@@ -67,60 +31,75 @@ func configureSwarm(p Provisioner, swarmOptions swarm.Options, authOptions auth.
...
-`
+		//Discovery must be at end of command
+		cmdMaster = append(cmdMaster, swarmOptions.Discovery)
 
-	if swarmOptions.Master {
-		log.Debug("Launching swarm master")
-		if err := runSwarmCommandFromTemplate(p, swarmMasterCmdTemplate, swarmCmdContext); err != nil {
+		hostBind := fmt.Sprintf("%s:%s", dockerDir, dockerDir)
+		masterHostConfig := dockerclient.HostConfig{
+			RestartPolicy: dockerclient.RestartPolicy{"Always", 0},
```
Comment (nathanleclaire) [inline libmachine/provision/configure_swarm.go]: `go vet` is complaining that you did not say which struct literal field `"Always"` and `0` are meant to belong to here.

**Comment by kunalkushwaha** [comment]
Ahh! It required field name explicitly.

This is new learning for me :). Gotools are really cool, they also teach indirectly :)

File: libmachine/provision/configure_swarm.go [hunk 2]
```diff
@@ -67,60 +31,79 @@ func configureSwarm(p Provisioner, swarmOptions swarm.Options, authOptions auth.
...
+		Cmd: []string{
+			"join",
+			"--advertise",
+			advertiseInfo,
+			swarmOptions.Discovery,
+		},
+		HostConfig: workerHostConfig,
+	}
+
+	err = CreateContainer(dockerClient, swarmWorkerConfig, "swarm-agent")
```
Comment (nathanleclaire) [inline libmachine/provision/configure_swarm.go]: If you want you can put this on just one line like so:

```golang
if err := CreateContainer(dockerClient, swarmWorkerConfig, "swarm-agent"); err != nil {
    return err
}
```

File: libmachine/provision/configure_swarm.go [hunk 3]
```diff
@@ -67,60 +31,78 @@ func configureSwarm(p Provisioner, swarmOptions swarm.Options, authOptions auth.
...
+		if err != nil {
 			return err
 		}
+
+	}
+
+	workerHostConfig := dockerclient.HostConfig{
+		RestartPolicy: dockerclient.RestartPolicy{
+			Name:              "Always",
+			MaximumRetryCount: 0},
```
Comment (nathanleclaire) [inline libmachine/provision/configure_swarm.go]: Nit: can you please put this `}` on the next line?

File: libmachine/provision/configure_swarm.go [hunk 4]
```diff
@@ -67,60 +31,78 @@ func configureSwarm(p Provisioner, swarmOptions swarm.Options, authOptions auth.
...
-	if _, err := p.SSHCommand(fmt.Sprintf("sudo docker pull %s", swarmOptions.Image)); err != nil {
-		return err
-	}
+		cmd := fmt.Sprintf("manage --tlsverify --tlscacert=%s --tlscert=%s --tlskey=%s -H %s --strategy %s --advertise %s",
+			authOptions.CaCertRemotePath,
+			authOptions.ServerCertRemotePath,
+			authOptions.ServerKeyRemotePath,
+			swarmOptions.Host,
+			swarmOptions.Strategy,
+			advertiseInfo)
```
Comment (nathanleclaire) [inline libmachine/provision/configure_swarm.go]: Nit: Can you please put this on next line?

**Comment by nathanleclaire** [comment]
LGTM

**Comment by dgageot** [comment]
LGTM

**Comment by nathanleclaire** [comment]
:+1:

**Pull request merged**